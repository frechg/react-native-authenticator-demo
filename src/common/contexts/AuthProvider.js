import React from 'react'
import * as SecureStore from 'expo-secure-store';
import { initialAuthState, authStateReducer } from '../authStateReducer';
import * as Api from '../../services/auth';

export const AuthContext = React.createContext();

export const AuthProvider = (props) => {
  const [authState, authDispatch] = React.useReducer(authStateReducer, initialAuthState);

  const getAuthState = async () => {
    try {
      const token = await SecureStore.getItemAsync('authToken');
      const username = await SecureStore.getItemAsync('username');
      authDispatch({type: 'RESTORE_TOKEN', token: token, username: username});
    } catch (error) {
      throw new Error(error);
    }
  };

  const signIn = async (data) => {
    await Api.signIn(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        throw new Error('Email or password are incorrect.');
      }
    })
  };

  const signUp = async (data) => {
    await Api.signUp(data)
    .then(async (response) => {
      if (response.ok) {
        const jsonResponse = await response.json();
        await SecureStore.setItemAsync('authToken', jsonResponse.token);
        await SecureStore.setItemAsync('username', jsonResponse.username);
        authDispatch({type: 'SIGN_IN', token: jsonResponse.token, username: jsonResponse.username});
      } else {
        const jsonResponse = await response.json();
        throw new Error(jsonResponse.errors[0])
      }
    });
  };

  const signOut = async () => {
    try {
      await SecureStore.deleteItemAsync('authToken');
      await SecureStore.deleteItemAsync('username');
      authDispatch({type: 'SIGN_OUT'});
    } catch (error) {
      throw new Error(error);
    }
  };

  const contextValue = React.useMemo(() => {
    return {authState, getAuthState, signUp, signIn, signOut};
  }, [authState]);

  return (
    <AuthContext.Provider value={contextValue}>
      {props.children}
    </AuthContext.Provider>
  );
};