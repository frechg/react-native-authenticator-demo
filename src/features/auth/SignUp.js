import React from 'react';
import {
    Text,
    View,
    TouchableWithoutFeedback,
    Keyboard,
    KeyboardAvoidingView,
    Platform
} from 'react-native';
import { BasicTextInput } from '../../common/components/BasicTextInput';
import { BasicSubmitButton } from '../../common/components/BasicSubmitButton';
import { styles, formStyles } from '../../common/styles';
import { AuthContext } from '../../common/contexts/AuthProvider';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';

export const SignUp = ({ navigation }) => {
  const { signUp } = React.useContext(AuthContext);

  const validation = Yup.object({
    username: Yup.string().required('Required'),
    email: Yup.string().email('Invalid email').required('Required'),
    password: Yup.string().required('Required'),
  });

  const handleSubmit = async (values) => {
    await signUp(values);
  }

  return (
    <KeyboardAvoidingView 
      behavior={Platform.OS === "ios" ? "padding" : "height"}
      style={styles.container}
    >
      <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
        <View style={formStyles.formWrapper}>
          <Formik
            initialValues={{username: '', email: '', password: ''}}
            validationSchema={validation}
            onSubmit={async (values, {setSubmitting, setStatus}) => {
              await handleSubmit(values).catch((error) => {
                setStatus(error.message)
                setSubmitting(false);
              });
              setSubmitting(false);
            }}
          >
            {({status}) => (
              <View>
                { status && <Text style={formStyles.formError}>{status}</Text> }
                <Field
                  component={BasicTextInput}
                  name="username"
                  placeholder="Username"
                />
                <Field
                  component={BasicTextInput}
                  name="email"
                  placeholder="Email"
                  keyboardType='email-address'
                />
                <Field
                  component={BasicTextInput}
                  name="password"
                  placeholder="Password"
                  secureTextEntry={true}
                />
                <BasicSubmitButton title='Sign Up' />
              </View>
            )}
          </Formik>
        </View>
      </TouchableWithoutFeedback>
    </KeyboardAvoidingView>
  );
}