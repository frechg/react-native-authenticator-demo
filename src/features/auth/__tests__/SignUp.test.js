import * as React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react-native';

import { SignUp } from '../SignUp';
import { AuthContext } from '../../../common/contexts/AuthProvider';

describe('<SignUp />', () => {
  test('Renders SignUp screen', () => {
    const signUp = jest.fn();
    render(
      <AuthContext.Provider value={{signUp}}>
        <SignUp />
      </AuthContext.Provider>
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('User submits form with valid account data', () => {
    const signUp = jest.fn();
    render(
      <AuthContext.Provider value={{signUp}}>
        <SignUp />
      </AuthContext.Provider>
    );

    // Setup account data
    const username = 'person';
    const email = 'person@email.com';
    const password = 'testingpass';

    // Enter and submit account data
    fireEvent.changeText(screen.getByPlaceholderText('Username'), username);
    fireEvent.changeText(screen.getByPlaceholderText('Email'), email);
    fireEvent.changeText(screen.getByPlaceholderText('Password'), password);
    fireEvent.press(screen.getByText('Sign Up'));

    // Expect account data to be submitted
    expect(signUp).toHaveBeenCalledWith({email: email, password: password, username: username});
  });

  test('User submits form without username', () => {
    const signUp = jest.fn();
    render(
      <AuthContext.Provider value={{signUp}}>
        <SignUp />
      </AuthContext.Provider>
    );

    // Enter and submit account data
    fireEvent.changeText(screen.getByPlaceholderText('Email'), 'person@email.com');
    fireEvent.changeText(screen.getByPlaceholderText('Password'), 'testingpass');
    fireEvent.press(screen.getByText('Sign Up'));

    // Expect account data to be submitted
    expect(screen.getByText('Username is required')).toBeTruthy();
  });

  test('User submits form without email', () => {
    const signUp = jest.fn();
    render(
      <AuthContext.Provider value={{signUp}}>
        <SignUp />
      </AuthContext.Provider>
    );

    // Enter and submit account data
    fireEvent.changeText(screen.getByPlaceholderText('Username'), 'Person');
    fireEvent.changeText(screen.getByPlaceholderText('Password'), 'testingpass');
    fireEvent.press(screen.getByText('Sign Up'));

    // Expect account data to be submitted
    expect(screen.getByText('Email is required')).toBeTruthy();
  });

  test('User submits form without password', () => {
    const signUp = jest.fn();
    render(
      <AuthContext.Provider value={{signUp}}>
        <SignUp />
      </AuthContext.Provider>
    );

    // Enter and submit account data
    fireEvent.changeText(screen.getByPlaceholderText('Username'), 'Person');
    fireEvent.changeText(screen.getByPlaceholderText('Email'), 'person@example.com');
    fireEvent.press(screen.getByText('Sign Up'));

    // Expect account data to be submitted
    expect(screen.getByText('Password is required')).toBeTruthy();
  });

  test('User sees error when sign up request fails', async () => {
    const errorMessage = 'Sign up failed';
    const signUp = jest
      .fn()
      .mockRejectedValue(new Error(errorMessage));
    
    render(
      <AuthContext.Provider value={{signUp}}>
        <SignUp />
      </AuthContext.Provider>
    );

    await waitFor(() => {
      // Enter and submit account data
      fireEvent.changeText(screen.getByPlaceholderText('Username'), 'Person');
      fireEvent.changeText(screen.getByPlaceholderText('Email'), 'person@example.com');
      fireEvent.changeText(screen.getByPlaceholderText('Password'), 'testingpass');
      fireEvent.press(screen.getByText('Sign Up'));

      // Expect account data to be submitted
      expect(screen.getByText(errorMessage)).toBeTruthy();
    });
  });
});