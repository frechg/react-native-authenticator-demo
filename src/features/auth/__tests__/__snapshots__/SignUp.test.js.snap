// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> Renders SignUp screen 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "rgb(245,245,245)",
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "padding": 32,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "marginBottom": 24,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(150,150,150)",
            "fontWeight": "bold",
            "marginBottom": 6,
            "textTransform": "uppercase",
          }
        }
      >
        Username
      </Text>
      <TextInput
        name="username"
        onChangeText={[Function]}
        placeholder="Username"
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "rgb(120,120,120)",
            "borderRadius": 6,
            "borderWidth": 1,
            "fontSize": 16,
            "padding": 12,
            "width": "100%",
          }
        }
        value=""
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "marginBottom": 24,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(150,150,150)",
            "fontWeight": "bold",
            "marginBottom": 6,
            "textTransform": "uppercase",
          }
        }
      >
        Email
      </Text>
      <TextInput
        keyboardType="email-address"
        name="email"
        onChangeText={[Function]}
        placeholder="Email"
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "rgb(120,120,120)",
            "borderRadius": 6,
            "borderWidth": 1,
            "fontSize": 16,
            "padding": 12,
            "width": "100%",
          }
        }
        value=""
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "marginBottom": 24,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(150,150,150)",
            "fontWeight": "bold",
            "marginBottom": 6,
            "textTransform": "uppercase",
          }
        }
      >
        Password
      </Text>
      <TextInput
        name="password"
        onChangeText={[Function]}
        placeholder="Password"
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "rgb(120,120,120)",
            "borderRadius": 6,
            "borderWidth": 1,
            "fontSize": 16,
            "padding": 12,
            "width": "100%",
          }
        }
        value=""
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#25273d",
          "borderRadius": 6,
          "opacity": 1,
          "padding": 12,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 18,
          }
        }
      >
        Sign Up
      </Text>
    </View>
  </View>
</View>
`;
